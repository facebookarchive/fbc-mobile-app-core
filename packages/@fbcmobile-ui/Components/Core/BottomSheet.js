/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

import * as React from 'react';
import Colors from '@fbcmobile/ui/Theme/Colors';
import RBSheet from 'react-native-raw-bottom-sheet';
import {useEffect, useRef} from 'react';

type Props = {
  +children: ?React.Node,
  +style?: ViewStyleProp,
  +title?: string,
  +height?: number,
  +isOpen?: boolean,
  +closeOnPressMask?: ?boolean,
  +closeOnPressBack?: ?boolean,
  +onClose?: () => void,
};

type RNSheetType = {
  open: () => void,
  close: () => void,
};

/**
 * Usage:
 * <BottomSheet
 *   isOpen={bottomSheetVisible}
 *   onClose={() => setBottomSheetVisible(false)}
 *   items={
 *     <View>
 *       ...
 *     </View>
 *   }
 * />
 */
const BottomSheet = ({
  isOpen = false,
  closeOnPressMask = false,
  closeOnPressBack = false,
  height = 250,
  children,
  onClose,
}: Props) => {
  const refRBSheet = useRef<?RNSheetType>(null);

  useEffect(() => {
    if (isOpen) {
      refRBSheet.current && refRBSheet.current.open();
    } else {
      refRBSheet.current && refRBSheet.current.close();
    }
  }, [isOpen]);

  return (
    <RBSheet
      ref={refRBSheet}
      height={height}
      duration={150}
      closeOnDragDown={false}
      closeOnPressMask={closeOnPressMask}
      closeOnPressBack={closeOnPressBack}
      onClose={onClose}
      customStyles={{
        draggableIcon: {width: 0, height: 0, margin: 0},
        container: {
          borderTopLeftRadius: 10,
          borderTopRightRadius: 10,
        },
        wrapper: {
          backgroundColor: Colors.TransparentBlack,
        },
      }}>
      {children}
    </RBSheet>
  );
};

export default BottomSheet;
