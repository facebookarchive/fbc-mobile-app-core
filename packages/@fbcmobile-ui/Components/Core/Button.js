/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

import * as React from 'react';
import Text from '@fbcmobile/ui/Components/Core/Text';
import {Colors} from '@fbcmobile/ui/Theme';
import {StyleSheet, TouchableOpacity} from 'react-native';

type ButtonVariant = 'primary' | 'plain' | 'secondary' | 'text' | 'warning';
type TextAlign = 'left' | 'center';

type Props = {
  +children: React.Node,
  +variant?: ButtonVariant,
  +textAlign?: TextAlign,
  +fontSize?: number,
  +onPress?: void | (() => void | Promise<void>),
  +disabled?: boolean,
  +style?: ViewStyleProp,
  +rightIcon?: React.Node,
};

const Button = ({
  children,
  onPress,
  disabled = false,
  variant = 'primary',
  style,
  textAlign = 'center',
  fontSize = 16,
  rightIcon,
}: Props) => {
  return (
    <TouchableOpacity
      style={StyleSheet.compose(
        [
          styles.root,
          styles[`${variant}Variant`],
          disabled ? styles[`${variant}Disabled`] : null,
        ],
        style,
      )}
      onPress={onPress}
      disabled={disabled}>
      <Text
        variant="h3"
        weight="bold"
        style={[
          styles.text,
          styles[`${variant}Text`],
          {fontSize: fontSize},
          disabled
            ? variant === 'text'
              ? styles.disabledTextVariant
              : styles.disabledText
            : null,
          styles[`${textAlign}TextAlign`],
        ]}>
        {children}
      </Text>
      {rightIcon}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  root: {
    paddingVertical: 10.5,
    paddingHorizontal: 12,
    borderRadius: 4,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  primaryVariant: {
    backgroundColor: Colors.Blue,
  },
  primaryDisabled: {
    backgroundColor: Colors.DisabledBackground,
  },
  secondaryVariant: {
    backgroundColor: Colors.BlueLight,
  },
  secondaryDisabled: {
    backgroundColor: Colors.DisabledBackground,
  },
  plainVariant: {
    backgroundColor: Colors.White,
    borderWidth: 1,
    borderColor: Colors.BlueGray,
  },
  plainDisabled: {
    backgroundColor: Colors.White,
    borderColor: Colors.DisabledText,
  },
  warningVariant: {
    backgroundColor: Colors.RedBackground,
  },
  warningDisabled: {
    backgroundColor: Colors.DisabledBackground,
  },
  textText: {
    color: Colors.Blue,
  },
  disabledTextVariant: {
    color: Colors.BlueTransparent,
  },
  disabledText: {
    color: Colors.DisabledText,
  },
  primaryText: {
    color: Colors.White,
  },
  secondaryText: {
    color: Colors.Blue,
  },
  plainText: {
    color: Colors.BlueGray,
  },
  warningText: {
    color: Colors.BrightRed,
  },
  leftTextAlign: {
    textAlign: 'left',
  },
  centerTextAlign: {
    textAlign: 'center',
  },
  text: {
    display: 'flex',
    flexGrow: 1,
    flexShrink: 1,
  },
});

export default Button;
