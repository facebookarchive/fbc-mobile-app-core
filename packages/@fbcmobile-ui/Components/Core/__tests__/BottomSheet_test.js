/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import BottomSheet from '../BottomSheet';
import Button from '@fbcmobile/ui/Components/Core/Button';
import React from 'react';
import renderer from 'react-test-renderer';

describe('BottomSheet Tests', () => {
  test('BottomSheet renders correctly', () => {
    global.managedClock(() => {
      const tree = renderer.create(
        <BottomSheet>
          <Button>Test</Button>
        </BottomSheet>,
      );
      const data = tree.toJSON();
      expect(data).toMatchSnapshot();
    });
  });
});
