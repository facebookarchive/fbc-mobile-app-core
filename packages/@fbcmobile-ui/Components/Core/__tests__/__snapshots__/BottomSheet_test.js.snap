// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet Tests BottomSheet renders correctly 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "portrait-upside-down",
      "landscape",
      "landscape-left",
      "landscape-right",
    ]
  }
  transparent={true}
  visible={false}
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Array [
          Object {
            "backgroundColor": "#00000077",
            "flex": 1,
          },
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
          },
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "opacity": 1,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderTopLeftRadius": 10,
          "borderTopRightRadius": 10,
          "height": 0,
          "overflow": "hidden",
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1877F2",
            "borderRadius": 4,
            "display": "flex",
            "flexDirection": "row",
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 10.5,
          }
        }
      >
        <Text
          style={
            Array [
              Array [
                Object {
                  "color": "#484848",
                  "fontSize": 16,
                  "fontWeight": "500",
                  "lineHeight": 20,
                },
                null,
                Object {
                  "fontFamily": "Roboto-Bold",
                },
                Object {
                  "textAlign": "left",
                },
              ],
              Array [
                Object {
                  "display": "flex",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                Object {
                  "color": "white",
                },
                Object {
                  "fontSize": 16,
                },
                null,
                Object {
                  "textAlign": "center",
                },
              ],
            ]
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
